Task to auto deactivate removed sso users

jingcheng16 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-15233
Tech spec: https://docs.google.com/document/d/1LOF1lCl3wTsEQkEEsHtZi3u4N6kvjXaLv-Lu6e9aYjo/edit#heading=h.ll7lzqwsaw9w

This task will iterate through all `IdentityProvider` objects where `enable_user_deactivation` is True. For now we will also filter by `idp_type` equal to `IdentityProviderType.AZURE_AD`. 

This task will query the [Microsoft Graph Users API](https://learn.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0&tabs=http#code-try-1) to fetch a list of all users that are members of the Azure AD Tenant that uses SSO to grant access to CommCare HQ. This task will fetch a list of usernames (email addresses) based on the userPrincipalName field returned by the Microsoft Graph Users API. It will then fetch a list of active WebUser usernames that are members of project spaces associated with the BillingAccount owner of the IdentityProvider. This task will then compare the two lists.

If a user is present in the list of active Web Users on HQ but is not present in the list returned by the Microsoft API query, we will deactivate that user by setting the is_active field on the WebUser to False. Any users that are in the “SSO Exempt” users list should not be automatically deactivated.

It is essential to handle errors in this task in a way that communicates issues swiftly both to the support team and the enterprise admins. For instance, if we are unable to access the Microsoft Graph API, we should raise the error to Sentry and send an email to the enterprise admins that we are having difficulty connecting with the Microsoft Graph API.

Additionally, if the Graph Users API returns an empty list of users, we should consider that an errored state and hold off from automatically deactivating Web Users if the number of active Web Users (that are not SSO Exempt) is greater than 3. This error should also be raised in an email to the enterprise admins alerting that we have temporarily skipped automatic deactivation of Web Users because we are receiving an empty list of Users from Azure AD.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I realized this task can be dangerous, we want to make sure we won't accidentally deactivate users that should not be deactivated. I wrote tests for the task, but did I catch all the edge cases?

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

corehq.apps.sso.tests.test_tasks:TestAutoDeactivationTask

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA at this stage

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
New dependency msal required for sso

jingcheng16 committed:
Task to autodeactivate web user

jingcheng16 committed:
lint

jingcheng16 committed:
Add test for auto deactivate web user

jingcheng16 committed:
Call send_html_email_async with delay

jingcheng16 committed:
Try to fix CI builds fail

jingcheng16 committed:
the mock send should be called with delay

biyeun commented on corehq/apps/sso/models.py:
```diff
@@ -404,6 +410,56 @@ def get_required_identity_provider(cls, username):
+                f"{result.get('error_description')}, Correlation ID: {result.get('correlation_id')}"
+                raise Exception(error_message)
+        except Exception as e:
+            notify_exception(None, f"Failed to get members of the IDP. {str(e)}")
```
nit: Generally `Identity Provider` is `IdP` in the reference docs elsewhere, rather than with a capital `D`
```suggestion
            notify_exception(None, f"Failed to get members of the IdP. {str(e)}")
```

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +108,45 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                body,
+            )
+        else:
+            users_to_deactivate = [user for user in web_user_in_account if user not in idp_users
```
what about users that are members of the domains but do not have an email domain controlled by the Identity Provider?

for instance, what if your dimagi user was a member of the domain?

I would like to see tests written for this scenario too :)

jingcheng16 replied:
Hi Biyeun, thank you for raising this important point! Fixed here: fc918be0f096ef40ba4d021f5be75647d3df3a85

biyeun commented on requirements/sso-requirements.in:
```diff
@@ -1,3 +1,4 @@
 pyOpenSSL>=23.2.0  # for cryptography>=41
 cryptography>=41   # for security vulnerability
 python3-saml
+msal
```
```suggestion
msal

```

biyeun replied:
new line at end of file nit

jingcheng16 replied:
Thank you Biyeun! Fixed here: a86ca6cb8243f2bf98b40ab643db861df72e7099 (I should just commit suggestion)

millerdev commented on requirements/docs-requirements.in:
```diff
@@ -7,5 +7,6 @@ sphinxcontrib-django
 sphinx-rtd-theme
 
 pyOpenSSL # CI builds fail without OpenSSL
+msal # CI builds fail without OpenSSL
```
I think this should be added in base-requirements.in and not here. This is not a docs requirement.

Alternately move the import in models.py into `get_all_members_of_the_idp` so it does not need to be added here.

jingcheng16 replied:
Thank you Daniel! I go with your second option : 11b11607f135468f8e9c84e184969e073dcd9e09

jingcheng16 committed:
IdP not IDP

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
only import msal in a function and remove it from docs-requirements

jingcheng16 committed:
New line at the end of the file

jingcheng16 committed:
Exclude user whose email domain is not controlled by the IdP

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                # If the user is not returned by IdP and is not exempted from SSO
+                if user not in idp_users and user not in exempt_usernames:
+                    email_domain = get_email_domain_from_username(user)
+                    authenticated_email_domains = authenticated_domains.values_list('email_domain', flat=True)
```
does this need to happen within the loop? this only needs to be set once, yes?

jingcheng16 committed:
Take authenticated_email_domains out of the loop

biyeun commented on corehq/apps/sso/models.py:
```diff
@@ -404,6 +409,57 @@ def get_required_identity_provider(cls, username):
+            notify_exception(None, f"Failed to get members of the IdP. {str(e)}")
+
+            # Send email
+            subject = _("Issue Connecting to Microsoft Graph API")
```
```suggestion
            subject = _("SSO Remote User Management: Issue Connecting to Microsoft Graph API")
```

jingcheng16 committed:
Update text

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

biyeun commented on corehq/apps/sso/models.py:
```diff
@@ -404,6 +409,57 @@ def get_required_identity_provider(cls, username):
+
+            # Send email
+            subject = _("Issue Connecting to Microsoft Graph API")
+            body = f'There was an issue connecting to the Microsoft Graph API: {str(e)}'
```
we should have a similar template [like the email sent here](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/apps/sso/utils/context_helpers.py#L51-L55)

It should read something like...
```
Dear enterprise administrator,

There was an issue connecting to the Microsoft Graph API when running a task to deactivate SSO Web Users on CommCare HQ that have been removed from your Microsoft Entra ID Enterprise Application.

This was the error message:
{{ error }}

Please check the configuration of the Remote User Management settings in your Identity Provider.

If you have any questions, please don’t hesitate to contact {{ contact_email }}.
Thank you for your use and support of CommCare.

Best regards,

The CommCare HQ Team
```

biyeun replied:
Additionally, the send email portion of this task should probably be its own utility. likewise with the similar suggestion below

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+
+        if len(idp_users) == 0 and len(web_user_in_account) - len(exempt_usernames) > 3:
+            # Send email
+            subject = _("Temporarily skipped automatic deactivation of Web Users")
```
```suggestion
            subject = _("CommCare HQ Alert: Temporarily skipped automatic deactivation of SSO Web Users (Remote User Management)")
```

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+            # Send email
+            subject = _("Temporarily skipped automatic deactivation of Web Users")
+            recipient = idp.owner.enterprise_admin_emails
+            body = _("we have temporarily skipped automatic deactivation of Web Users because we are receiving "
```
Similar to the previous suggestion for updating the `body` text in `get_all_members_of_the_idp`

```
Dear enterprise administrator,

We have temporarily skipped automatic deactivate of SSO Web Users because we are receiving an empty list of users from your Microsoft Entra ID instance.

Please check the configuration of the Remote User Management settings in your Identity Provider.

If you have any questions, or if this issue persists, please don’t hesitate to contact {{ contact_email }}.
Thank you for your use and support of CommCare.

Best regards,

The CommCare HQ Team
```

biyeun replied:
this email bit can also be its own utility abstracted away

jingcheng16 replied:
Thank you Biyeun. Improved in 9128522fe0edb09107879367efc6cfb52478079b

biyeun commented on corehq/apps/sso/models.py:
```diff
@@ -404,6 +409,57 @@ def get_required_identity_provider(cls, username):
+                user_principal_names = [user["userPrincipalName"] for user in graph_data["value"]]
+                return user_principal_names
+            else:
+                error_message = f"Error acquiring token: {result.get('error')}, Description: "
```
```suggestion
                error_message = f"Error acquiring Microsoft Graph API token: {result.get('error')}, Description: "
```

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+        for domain in domains:
+            [web_user_in_account.add(user.username) for user in WebUser.by_domain(domain)]
+
+        # Get a list of SSO Exempt users
```
```suggestion
        # Get criteria for exempting usernames and email domains from the deactivation list
```

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+        enable_user_deactivation=True,
+        idp_type=IdentityProviderType.AZURE_AD
+    ).all():
+        # Fetch a list of users usernames that are members of the idp
```
```suggestion
        # Fetch a list of users usernames that are members of domains belonging to the Identity Provider's Billing Account owner
```

biyeun replied:
can the resulting list be its own utility?

jingcheng16 replied:
Hi Biyeun, this function will get all users of the idp, even though some of them might not be our hq web user

gherceg replied:
I would say that this comment isn't necessary here. The variable name and method name should be descriptive enough, and if someone wants more of an explanation, they can look at the docstring for the method  `get_all_members_of_the_idp`. So I'd suggest moving this comment to a docstring on the method.

jingcheng16 replied:
Go with Graham's suggestion in  9e871028b9474127c75e96287df074446cfd57ce

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                subject,
+                recipient,
+                body,
+            )
```
i generally would `return` here and remove the following `else` block, since the code in the `else` block is quite lengthy. Better for readability imo

jingcheng16 replied:
Thank you. Fixed in c2ec7bac32f420db122d6730ab5f2fa5678eb44c

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,53 @@ def send_idp_cert_expires_reminder_emails(num_days):
+        # Fetch a list of users usernames that are members of the idp
+        idp_users = idp.get_all_members_of_the_idp()
+        # Fetch a list of active WebUser usernames that are members of project spaces associated with the idp
+        domains = idp.owner.get_domains()
```
this and the following lines are very similar to 
https://github.com/dimagi/commcare-hq/blob/9f3b3414a3007c48d612d06c7ba3219fcdee1124/corehq/apps/accounting/tasks.py#L836-L839

possible to make a utility or instance method on `BillingAccount`?

jingcheng16 replied:
Thank you Biyeun. Refactored here: 9b27bcefe42dc5fd0fd9e46cca1920cb2a06751b

jingcheng16 committed:
Update text

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
Better error message

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
Update comment

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
lint

jingcheng16 committed:
Better readability

gherceg commented on corehq/apps/sso/tests/test_tasks.py:
```diff
@@ -133,3 +139,103 @@ def tearDown(self):
+        self.web_user_d = self._create_web_user('d@vaultwax.com')
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
+    def test_user_deactivation_logic(self, mock_get_all_members_of_the_idp):
```
nit: would be nice to see a more specific test name. What is it about the deactivation logic that you are testing? It looks like you are testing that web_user_c is deactivated if no longer part of the idp? So maybe `test_user_is_deactivated_if_not_member_of_idp`

jingcheng16 replied:
Good suggestion! Thank you! Refactored here: a1940ef363250d2c4d7b5ca6b1ee708fb4081e81

gherceg commented on corehq/apps/sso/tests/test_tasks.py:
```diff
@@ -133,3 +139,103 @@ def tearDown(self):
+        self.web_user_a = self._create_web_user('a@vaultwax.com')
+        self.web_user_b = self._create_web_user('b@vaultwax.com')
+        self.web_user_c = self._create_web_user('c@vaultwax.com')
+        self.web_user_d = self._create_web_user('d@vaultwax.com')
```
Is this user ever needed for any of these tests?

jingcheng16 replied:
It is needed because of `len(web_user_in_account) - len(exempt_usernames) > 3`. I can left a comment here: 28aaa49c3bf080454d2f36941b64d7a2f7d129e9

gherceg replied:
I see. Doesn't require any action, but the downside of setting up these users in setUp is that you can't easily test the scenario where auto-deactivation is _not_ skipped because  `len(web_user_in_account)_ - len(exempt_usernames) <= 3`. Is that fair to say?

jingcheng16 replied:
Yes that's fair. :face_holding_back_tears:

gherceg commented on corehq/apps/sso/tests/test_tasks.py:
```diff
@@ -133,3 +139,103 @@ def tearDown(self):
+        self.assertTrue(web_user_c.is_active)
+        mock_send.assert_called_once()
+
+    @patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
```
super nit: Since you are patching this for every test, setting this patcher up at the class level could help make each test a bit more concise and readable. That could just be moving the decorator to the class, but I like setting up the patcher in setUp (or setUpClass), because you avoid having the extra parameter in each test method. Something like this:
```
        idp_patcher = patch('corehq.apps.sso.models.IdentityProvider.get_all_members_of_the_idp')
        cls.mock_get_members_of_idp = idp_patcher.start()
        cls.addClassCleanup(idp_patcher.stop)
```

jingcheng16 replied:
Thank you! Love this suggestion. Refactored here: 080b22a5c40adab3718728ed3261e3ad3afccfdb

gherceg commented on corehq/apps/sso/tests/test_tasks.py:
```diff
@@ -133,3 +139,103 @@ def tearDown(self):
+        cls.account = generator.get_billing_account_for_idp()
+        cls.account.enterprise_admin_emails = ['test@vaultwax.com']
+        cls.account.save()
+        cls.domain = Domain.get_or_create_with_name("vaultwax-001", is_active=True)
```
Add `cls.addClassCleanup(cls.domain.delete)`

jingcheng16 replied:
THANK YOU!!! Added in 602208719e9ee591431876a37c68bc37f357aee6

gherceg reviewed (commented):
Looking good. Mostly minor comments with a few that need to be fixed before approving.

jingcheng16 committed:
Simplify exmept_usernames query logic

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

jingcheng16 committed:
lint

jingcheng16 committed:
Add comment to explain why web_user_d is required

jingcheng16 committed:
Refactored the process of gathering usernames from WebUsers across all domains within an account

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

jingcheng16 committed:
lint

jingcheng16 committed:
Code is self-explanatory so remove comment

jingcheng16 commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,55 @@ def send_idp_cert_expires_reminder_emails(num_days):
+
+        for user in web_user_in_account:
+            # If the user is not returned by IdP and is not exempted from SSO
+            if user not in idp_users and user not in exempt_usernames:
```
Thank you, removed in 2ea9229df77a001740a709ab3712175631ab586c

jingcheng16 committed:
Remove unnecessary comment

jingcheng16 commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,55 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                email_domain = get_email_domain_from_username(user)
+                # If the user's email domain is in the list of email domains controlled by IdP
+                if email_domain in authenticated_email_domains:
+                    users_to_deactivate.append(user)
```
Hi Graham, I removed my comment in: 2ffda903f204a989fd72babfa21ae34cd9d419d8... I was planning to explain why, then I felt it is redundant... we only want to deactivate user that is no longer in the organization of this IdP, so if a email domain is not controlled by the IdP, then we shouldn't deactivate it.

gherceg commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,55 @@ def send_idp_cert_expires_reminder_emails(num_days):
+        sso_exempt_users = UserExemptFromSingleSignOn.objects.filter(email_domain__in=authenticated_domains)
+        exempt_usernames = sso_exempt_users.values_list('username', flat=True)
+
+        if len(idp_users) == 0 and len(web_user_in_account) - len(exempt_usernames) > 3:
```
Another option is to encapsulate `len(web_users_in_account) - len(exempt_usernames) > 3` in a little helper where 3 is still hard coded (since it seems like this number isn't used elsewhere), but the method name and docstring provides a clear idea of what the helper is checking for/why we chose 3.

jingcheng16 replied:
Thank you both! Made a constant and left a comment here: e515ddada9fd16c2012217992ecb67383199a095

gherceg replied:
I offlined with Jing about this magic number some more because I'm intrigued. She linked [this thread](https://docs.google.com/document/d/1LOF1lCl3wTsEQkEEsHtZi3u4N6kvjXaLv-Lu6e9aYjo/edit) from the tech spec which was helpful, but it seems that conversation focused on the endpoint returning an empty list and not on this magic number. @biyeun you had said:
> we should consider that an errored state and hold off from automatically deactivating Web Users if the number of active Web Users (that are not SSO Exempt) is greater than 3

with your concern being incorrectly deactivating all web users for a project. I'm curious why it can't be the case that if we are in this errored state and there are _any_ active web users that are not SSO exempt, we should hold off on deactivating web users? I know it is late in the game to bring this up, but given it is a simple change, figured it is still worth asking. Thanks!

biyeun replied:
@gherceg I was trying to think of a case where it's likely an empty list is not an errored state, but perhaps we should always treat an empty list as a potentially errored state. thoughts?

jingcheng16 replied:
I prefer @gherceg 's proposal. @biyeun , do you mean always treat an empty list as a potentially errored state, even if there are no active users?

biyeun replied:
@gherceg the reason why I chose 3 in the spec, is that felt like more of an active state of using this feature rather than a testing state. I feel if there are more than 3 active SSO users then this feature is in active use, and we should be concerned about potentially deactivating all users if the list returns empty. However, in a testing state (3 users or below) we should respect an empty list because it's possible the list of users could be empty on the azure side. Thoughts? This was stated in the spec with caution around the proposal, but no one had responded during review.

gherceg replied:
Sorry for completely missing this thread! And yeah apologies again for bringing this up in the implementation stage and not the spec review stage.

So we are trying to differentiate between an error state, a testing state, and an active state. Given that it is hard to clearly identify a testing state, I think removing it would simplify this feature and lessen the burden on documenting this magic number of 3. If our policy is simply that when an empty list is returned from an IDP, our deactivation code will not execute and we will communicate that we have skipped this to the customer, that seems more straightforward.

But again, I'm bringing this up late and don't have all of the context, so if you feel that attempting to identify this testing state is worthwhile, feel free to move forward with this as is.

biyeun replied:
alright. @jingcheng16 let's remove the 3 user requirement and always treat an empty list as an errored state

jingcheng16 replied:
@biyeun and @gherceg , the threshold removed in 66b26a40d36868585cf193b5630dc43efa7fcf22

jingcheng16 committed:
Refactor: "user" to refer to WebUser objects, "username" to refer to the actual string property

jingcheng16 commented on corehq/apps/sso/tasks.py:
```diff
@@ -99,3 +109,55 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                    users_to_deactivate.append(user)
+
+        # Deactivate user that is not returned by Graph Users API
+        for username in users_to_deactivate:
```
Thank you! I should make it more clear. Refactored here: 9b31683ecfdbc7e470c55ec212bd38f7b64a1244

jingcheng16 committed:
Refactor: a more specific test name

jingcheng16 committed:
Refactor: set the patcher up at the class level

jingcheng16 committed:
Add class cleanup

jingcheng16 committed:
lint

jingcheng16 committed:
Refactor: create method on BillingAccount and call it in tasks

jingcheng16 committed:
Utility function and new template for api connection issue email

jingcheng16 committed:
wrong file name...

jingcheng16 committed:
Add baseurl as it is required by the template

jingcheng16 committed:
Add email context utility for deactivation skip notifaction

jingcheng16 committed:
handle error if no admin email

jingcheng16 committed:
Remove redundant comments for self-explanatory code

jingcheng16 committed:
Wrong intendation

jingcheng16 committed:
Have a constant for the minimum active user threshold

jingcheng16 committed:
Merge branch 'master' into jc/auto-deactivate-removed-sso-users

jingcheng16 committed:
Merge branch 'master' into jc/auto-deactivate-removed-sso-users

jingcheng16 committed:
Rename Azure to Entra

jingcheng16 committed:
Merge branch 'master' into jc/auto-deactivate-removed-sso-users

jingcheng16 committed:
Remove the threshold for skip auto deactivation

biyeun commented on corehq/apps/sso/templates/sso/email/microsoft_graph_api_connection_issue_notification.txt:
```diff
@@ -0,0 +1,20 @@
+This was the error message:
+{{ error }}
+
+Please check the configuration of the Remote User Management settings in your Identity Provider.
```
```suggestion
Please check the configuration for Remote User Management in your Identity Provider settings.
```

biyeun commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.txt:
```diff
@@ -0,0 +1,17 @@
+{% blocktrans %}
+Dear enterprise administrator,
+
+We have temporarily skipped automatic deactivate of SSO Web Users
```
```suggestion
We have temporarily skipped automatic deactivation of SSO-managed Web Users
```

biyeun commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.html:
```diff
@@ -0,0 +1,34 @@
+
+<p>
+  {% blocktrans %}
+    Please check the configuration of the Remote User Management settings in your Identity Provider.
```
```suggestion
    Please check the configuration for Remote User Management in your Identity Provider settings.
```

biyeun reviewed (commented):
made some text update suggestions. but will ✅ once those are done. Thanks!

gherceg commented on corehq/apps/sso/models.py:
```diff
@@ -418,6 +426,71 @@ def get_required_identity_provider(cls, username):
+            else:
+                error_message = f"Error acquiring Microsoft Graph API token: {result.get('error')}, Description: "
+                f"{result.get('error_description')}, Correlation ID: {result.get('correlation_id')}"
+                raise Exception(error_message)
+        except Exception as e:
```
Typically we try to avoid raising general exceptions (Exception base class), since we then have to catch general exceptions (which we also try to avoid). This might be a good use case for a custom exception class.

gherceg commented on corehq/apps/sso/models.py:
```diff
@@ -418,6 +426,71 @@ def get_required_identity_provider(cls, username):
+                            "send_to": send_to,
+                        }
+                    )
+            except Exception as exc:
```
Is there a specific exception you expect to be raised from this? It looks like fairly straightforward code that triggers celery tasks.

gherceg commented on corehq/apps/sso/tasks.py:
```diff
@@ -110,6 +119,64 @@ def send_idp_cert_expires_reminder_emails(num_days):
+                            "send_to": send_to,
+                        }
+                    )
+            except Exception as exc:
```
Same comment as above, if there is a specific exception you know could be raised, catch it, otherwise this try catch is likely not needed.

gherceg commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.html:
```diff
@@ -0,0 +1,34 @@
+<p>
+  {% blocktrans %}The CommCare HQ Team{% endblocktrans %}<br />
+  {{ base_url }}
+</p>
```
nit: missing newline

gherceg commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.html:
```diff
@@ -0,0 +1,34 @@
+
+<p>
+  {% blocktrans %}
+    We have temporarily skipped automatic deactivate of SSO Web Users
+    because we are receiving an empty list of users from your Microsoft Entra ID instance.
```
```suggestion
    We have temporarily skipped automatic deactivation of SSO Web Users
    because we received an empty list of users from your Microsoft Entra ID instance.
```

gherceg commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.txt:
```diff
@@ -0,0 +1,17 @@
+{% blocktrans %}
+Dear enterprise administrator,
+
+We have temporarily skipped automatic deactivate of SSO Web Users
+because we are receiving an empty list of users from your Microsoft Entra ID instance.
```
```suggestion
We have temporarily skipped automatic deactivation of SSO Web Users
because we received an empty list of users from your Microsoft Entra ID instance.
```

gherceg commented on corehq/apps/sso/utils/context_helpers.py:
```diff
@@ -66,6 +66,56 @@ def get_idp_cert_expiration_email_context(idp):
 
 
+def get_graph_api_connection_issue_email_context(idp, error):
+    subject = _("CommCare HQ Alert: SSO Remote User Management, Issue Connecting to Microsoft Graph API")
```
nit
```suggestion
    subject = _("CommCare HQ Alert: SSO Remote User Management - Issue Connecting to Microsoft Graph API")
```

gherceg reviewed (commented):
Would like to see the changes to `get_web_users` I mentioned before approving, but otherwise it looks great!

jingcheng16 committed:
Text update

Co-authored-by: Biyeun <bbuczyk@dimagi.com>
Co-authored-by: Graham Herceg <gherceg@dimagi.com>

biyeun commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.html:
```diff
@@ -0,0 +1,34 @@
+<p>
+  {% blocktrans %}
+    We have temporarily skipped automatic deactivation of SSO-managed Web Users
+    because we are receiving an empty list of users from your Microsoft Entra ID instance.
```
I like @gherceg's suggestion here
```suggestion
    because we received an empty list of users from your Microsoft Entra ID instance.
```

biyeun commented on corehq/apps/sso/templates/sso/email/sso_deactivation_skip_notification.txt:
```diff
@@ -0,0 +1,17 @@
+Dear enterprise administrator,
+
+We have temporarily skipped automatic deactivation of SSO-managed Web Users
+because we are receiving an empty list of users from your Microsoft Entra ID instance.
```
```suggestion
because we received an empty list of users from your Microsoft Entra ID instance.
```

jingcheng16 committed:
Update text

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
Refactor: better query web users by domain

jingcheng16 commented on corehq/apps/accounting/models.py:
```diff
@@ -603,6 +603,25 @@ def _send_autopay_card_added_email(self, domain):
             use_domain_gateway=True,
         )
 
+    def get_web_users(self, info_type='id'):
```
Thank you. Refactored in : 0473337f1290ffc09fc1435f61a66e90556ab8a1

jingcheng16 committed:
Remove unnecessary try catch

jingcheng16 committed:
One auto deactivation skipped email for different failure reason

jingcheng16 committed:
Miss one parameter

biyeun commented on corehq/apps/sso/models.py:
```diff
@@ -427,69 +424,10 @@ def get_required_identity_provider(cls, username):
-            except Exception as exc:
-                notify_exception(None, f"Failed to send Microsoft Graph API connection issue"
-                                 f" for IdP {self.name}: {exc!s}", exc_info=True)
+        if self.idp_type == IdentityProviderType.ENTRA_ID:
```
thanks for this change, much better!

jingcheng16 committed:
Return user in an application

jingcheng16 committed:
Add a new line at the end of the file

gherceg commented on corehq/apps/sso/utils/context_helpers.py:
```diff
@@ -81,7 +81,7 @@ def get_graph_api_connection_issue_email_context(idp, error):
         "subject": subject,
         "from": _(f"Dimagi CommCare Accounts <{settings.ACCOUNTS_EMAIL}>"),
-        "to": idp.owner.enterprise_admin_emails,
+        "to": idp.owner.enterprise_admin_emails or idp.owner.dimagi_contact or settings.ACCOUNT_EMAIL,
```
```suggestion
        "to": idp.owner.enterprise_admin_emails or [idp.owner.dimagi_contact] or [settings.ACCOUNT_EMAIL],
```

gherceg commented on corehq/apps/sso/utils/context_helpers.py:
```diff
@@ -106,7 +106,7 @@ def get_sso_deactivation_skip_email_context(idp):
         "subject": subject,
         "from": _(f"Dimagi CommCare Accounts <{settings.ACCOUNTS_EMAIL}>"),
-        "to": idp.owner.enterprise_admin_emails,
+        "to": idp.owner.enterprise_admin_emails or idp.owner.dimagi_contact or settings.ACCOUNT_EMAIL,
```
```suggestion
        "to": idp.owner.enterprise_admin_emails or [idp.owner.dimagi_contact] or [settings.ACCOUNT_EMAIL],
```

gherceg commented on corehq/apps/sso/utils/entra.py:
```diff
@@ -0,0 +1,48 @@
+    )
+    # looks up a token from cache
+    result = app.acquire_token_silent(config["scope"], account=None)
+    # try:
```
remove this comment?

jingcheng16 replied:
Thank you for catching this! Removed in 674be9f4176fe1f39c8b04348be97cd4974203f4

jingcheng16 committed:
email_context's "to" only accept iterable

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

jingcheng16 committed:
Remove wrong comment

biyeun commented on corehq/apps/sso/tests/test_tasks.py:
```diff
@@ -298,3 +302,101 @@ def _create_user(self, username):
+        self.assertTrue(web_user_c.is_active)
+        mock_send.assert_called_once()
+
+    def test_deactivation_skip_members_of_the_domains_but_not_have_an_email_domain_controlled_by_the_IdP(self):
```
```suggestion
    def test_deactivation_skip_members_of_the_domains_but_not_have_an_email_domain_controlled_by_the_idp(self):
```
very nit suggestion

biyeun reviewed (approved):
great! thank you! 🎉

gherceg reviewed (approved):
Thanks for your patience on this one Jing. Great job!

jingcheng16 committed:
rename IdP to idp

Co-authored-by: Biyeun <bbuczyk@dimagi.com>

jingcheng16 committed:
Restructure EntraVerificationFailed

The result returned looks like this: {'error': 'invalid_client',
 'error_description': "AADSTS7000215: Invalid client secret provided. Ensure the secret being sent in the request is the client secret value, not the client secret ID, for a secret added to app 'f09df325-2ca9-4756-93d4-14934194ae35'. Trace ID: fa552a4f-bb3b-45e0-8432-792489cf1e00 Correlation ID: 379dcf42-fc82-4a36-bce0-d8823be8046b Timestamp: 2024-04-17 14:26:07Z",
 'error_codes': [7000215],
 'timestamp': '2024-04-17 14:26:07Z',
 'trace_id': 'fa552a4f-bb3b-45e0-8432-792489cf1e00',
 'correlation_id': '379dcf42-fc82-4a36-bce0-d8823be8046b',
 'error_uri': 'https://login.microsoftonline.com/error?code=7000215'}

jingcheng16 committed:
Merge branch 'master' into jc/auto-deactivate-removed-sso-users

jingcheng16 committed:
Catch general exception so we still process other idps

jingcheng16 committed:
Fix keyword argument

jingcheng16 committed:
Raise error for permission denied

The batch_result will look like: {'responses': [{'id': '3',
   'status': 403,
   'headers': {'Cache-Control': 'no-cache',
    'x-ms-resource-unit': '1',
    'Content-Type': 'application/json'},
   'body': {'error': {'code': 'Authorization_RequestDenied',
     'message': 'Insufficient privileges to complete the operation.',
     'innerError': {'date': '2024-04-24T15:36:38',
      'request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd',
      'client-request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd'}}}},
  {'id': '1',
   'status': 403,
   'headers': {'Cache-Control': 'no-cache',
    'x-ms-resource-unit': '1',
    'Content-Type': 'application/json'},
   'body': {'error': {'code': 'Authorization_RequestDenied',
     'message': 'Insufficient privileges to complete the operation.',
     'innerError': {'date': '2024-04-24T15:36:38',
      'request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd',
      'client-request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd'}}}},
  {'id': '2',
   'status': 403,
   'headers': {'Cache-Control': 'no-cache',
    'x-ms-resource-unit': '1',
    'Content-Type': 'application/json'},
   'body': {'error': {'code': 'Authorization_RequestDenied',
     'message': 'Insufficient privileges to complete the operation.',
     'innerError': {'date': '2024-04-24T15:36:38',
      'request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd',
      'client-request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd'}}}},
  {'id': '4',
   'status': 403,
   'headers': {'Cache-Control': 'no-cache',
    'x-ms-resource-unit': '1',
    'Content-Type': 'application/json'},
   'body': {'error': {'code': 'Authorization_RequestDenied',
     'message': 'Insufficient privileges to complete the operation.',
     'innerError': {'date': '2024-04-24T15:36:38',
      'request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd',
      'client-request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd'}}}},
  {'id': '0',
   'status': 403,
   'headers': {'Cache-Control': 'no-cache',
    'x-ms-resource-unit': '1',
    'Content-Type': 'application/json'},
   'body': {'error': {'code': 'Authorization_RequestDenied',
     'message': 'Insufficient privileges to complete the operation.',
     'innerError': {'date': '2024-04-24T15:36:38',
      'request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd',
      'client-request-id': '30dbadf0-2abc-42aa-b838-280a99a477dd'}}}}]}

biyeun reviewed (changes requested):
retracting approval. 
The general error escape shouldn't be sent as an email. those errors we want to use `notify_exception` and not send the error message raw to the user

biyeun commented on corehq/apps/sso/tasks.py:
```diff
@@ -141,6 +141,12 @@ def auto_deactivate_removed_sso_users():
+            notify_exception(None, f"Failed to get members of the IdP. {str(e)}")
+            send_deactivation_skipped_email(idp=idp, failure_reason=MSGraphIssue.VERIFICATION_ERROR,
+                                            error="verification error",
+                                            error_description={str(e)})
```
We should avoid sending the raw error to the user, in the chance it contains sensitive information we didn't expect. `notify_exception` is fine and necessary, but for this we should say "unknown issue: please contact support"

biyeun replied:
thanks!

jingcheng16 replied:
Hi Biyeun, changed in 8871badc4af1022a6b2cfa5dccbc78483274bec9

jingcheng16 commented:
@biyeun 
The general exception might catch wrong configuration. 
```
    app = msal.ConfidentialClientApplication(
        config["client_id"], authority=config["authority"],
        client_credential=config["secret"],
    )
 ```
 This initialization can raise ValueError or general exception.

jingcheng16 committed:
Avoid sending raw error to user

biyeun reviewed (approved):
thank you!
